src/once_cell.rs

```diff
@@ -0,0 +1,800 @@
+    ///
+    /// let cell = OnceCell::from(42);
+    /// assert_eq!(cell.get(), Some(&42));
+    /// ```
+    fn from(value: T) -> Self {
+        Self {
+            active_initializers: Event::new(),
+            passive_waiters: Event::new(),
+            state: AtomicUsize::new(INITIALIZED),
+            value: UnsafeCell::new(MaybeUninit::new(value)),
+        }
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for OnceCell<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct Inner<'a, T>(&'a OnceCell<T>);
+
+        impl<T: fmt::Debug> fmt::Debug for Inner<'_, T> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                match self.0.state.load(Ordering::Acquire) {
+                    UNINITIALIZED => f.write_str("<uninitialized>"),
+                    INITIALIZING => f.write_str("<initializing>"),
+                    INITIALIZED => {
+                        // SAFETY: "value" is initialized.
+                        let value = unsafe { self.0.get_unchecked() };
+                        fmt::Debug::fmt(value, f)
+                    }
+                    _ => unreachable!("Invalid state"),
+                }
+            }
+        }
+
+        f.debug_tuple("OnceCell").field(&Inner(self)).finish()
+    }
+}
+
+/// Either return the result of a future now, or panic.
+fn now_or_never<T>(f: impl Future<Output = T>) -> T {
+    const NOOP_WAKER: RawWakerVTable = RawWakerVTable::new(clone, wake, wake_by_ref, drop);
+
+    unsafe fn wake(_: *const ()) {}
+    unsafe fn wake_by_ref(_: *const ()) {}
+    unsafe fn clone(_: *const ()) -> RawWaker {
+        RawWaker::new(ptr::null(), &NOOP_WAKER)
+    }
+    unsafe fn drop(_: *const ()) {}
+
+    futures_lite::pin!(f);
+    let waker = unsafe { Waker::from_raw(RawWaker::new(ptr::null(), &NOOP_WAKER)) };
+
+    // Poll the future exactly once.
+    let mut cx = Context::from_waker(&waker);
+
+    match f.poll(&mut cx) {
+        Poll::Ready(value) => value,
+        Poll::Pending => unreachable!("future not ready"),
+    }
+}
+
+/// The strategy for polling an `event_listener::EventListener`.
+trait Strategy {
+    /// Poll the event listener.
+    fn poll(&mut self, evl: EventListener, ctx: &mut Context<'_>) -> Result<(), EventListener>;
+}
+
+/// The strategy for blocking the current thread on an `EventListener`.
+struct Blocking;
+
+impl Strategy for Blocking {
+    fn poll(&mut self, evl: EventListener, _: &mut Context<'_>) -> Result<(), EventListener> {
+        evl.wait();
+        Ok(())
+    }
+}
+
+/// The strategy for polling an `EventListener` in an async context.
+struct NonBlocking;
+
+impl Strategy for NonBlocking {
+    fn poll(&mut self, mut evl: EventListener, ctx: &mut Context<'_>) -> Result<(), EventListener> {
+        match Pin::new(&mut evl).poll(ctx) {
+            Poll::Pending => Err(evl),
+            Poll::Ready(()) => Ok(()),
+        }
+    }
+}
+
+struct CallOnDrop<F: Fn()>(F);
```

Comment:
This doesn't seem to be used. :thinking:

Reply:
My bad, I made an original implementation that used that.